core/src/main/java/org/apache/iceberg/encryption/KeyMetadataDecoder.java

```diff
@@ -0,0 +1,145 @@
+    this.versionBuffer = new byte[1];
+  }
+
+  /**
+   * Adds an {@link org.apache.iceberg.Schema Iceberg schema} that can be used to decode buffers.
+   *
+   * @param writeSchema a schema to use when decoding buffers
+   */
+  public void addSchema(byte schemaVersion, org.apache.iceberg.Schema writeSchema) {
+    decoders.put(
+        schemaVersion,
+        new KeyMetadataDecoder.RawDecoder<>(
+            readSchema, AvroSchemaUtil.convert(writeSchema, "key_metadata")));
+  }
+
+  @Override
+  public D decode(InputStream stream, D reuse) throws IOException {
+    try {
+      if (!readFully(stream, versionBuffer)) {
+        throw new BadHeaderException("Failed to read version - no bytes in stream");
+      }
+    } catch (IOException e) {
+      throw new IOException("Failed to read the version byte", e);
+    }
+
+    byte writeSchemaVersion = versionBuffer[0];
+
+    RawDecoder<D> decoder = decoders.get(writeSchemaVersion);
+
+    if (decoder == null) {
+      throw new MissingSchemaException("Cannot resolve schema for version: " + writeSchemaVersion);
+    }
+
+    return decoder.decode(stream, reuse);
+  }
+
+  private static class RawDecoder<D> extends BaseDecoder<D> {
+    private final DatumReader<D> reader;
+    private BinaryDecoder binaryDecoder;
+
+    /**
+     * Creates a new {@link MessageDecoder} that constructs datum instances described by the {@link
+     * Schema readSchema}.
+     *
+     * <p>The {@code readSchema} is used for the expected schema and the {@code writeSchema} is the
+     * schema used to decode buffers. The {@code writeSchema} must be the schema that was used to
+     * encode all buffers decoded by this class.
+     *
+     * @param readSchema the schema used to construct datum instances
+     * @param writeSchema the schema used to decode buffers
+     */
+    private RawDecoder(org.apache.iceberg.Schema readSchema, Schema writeSchema) {
+      this.reader =
+          new ProjectionDatumReader<>(
+              avroSchema -> DataReader.create(readSchema, avroSchema),
+              readSchema,
+              ImmutableMap.of(),
+              null);
+      this.reader.setSchema(writeSchema);
+    }
+
+    @Override
+    public D decode(InputStream stream, D reuse) {
+      binaryDecoder = DecoderFactory.get().directBinaryDecoder(stream, binaryDecoder);
+      try {
+        return reader.read(reuse, binaryDecoder);
+      } catch (IOException e) {
+        throw new AvroRuntimeException("Decoding datum failed", e);
+      }
+    }
+  }
+
+  /**
+   * Reads a buffer from a stream, making multiple read calls if necessary.
+   *
+   * @param stream an InputStream to read from
+   * @param bytes a buffer
+   * @return true if the buffer is complete, false otherwise (stream ended)
+   * @throws IOException if there is an error while reading
+   */
+  @SuppressWarnings("checkstyle:InnerAssignment")
+  private boolean readFully(InputStream stream, byte[] bytes) throws IOException {
```

Comment:
Was this copied? I think we should use the one in `IOUtil`.

Reply:
Looks like there was one in IcebergDecoder. We should either move this to `IOUtil` or use one based on those methods.

Reply:
Yep, this is a copy. Removed per the discussion below (replaced with a simple stream.read)